rfcs/20210219-cache-compiled-gpu-kernels/README.md

```diff
@@ -0,0 +1,101 @@
+regardless of whether it's a cache hit or cache miss.
+
+This approach has a number of drawbacks which become critical:
+1. There is overhead on creating a kernel out of a binary per each primitive creation
+    * That time can vary significantly depending on architecture. For example creating
+    kernels from binary for batch normalization on Gen9 takes ~200us but on a new
+    architecture it may take ~1.5ms. The reason is likely not optimized software
+    stack for that architecture but it means that such an issue can occur for each
+    new architecture
+2. Unfortunately memory is context dependent and hence cannot be stored in the
+primitive. The scales must be copied to the memory and the memory then
+is stored in the oneDNN resource. The copy also involves map/unmap operations
+3. SYCL is trying to abstract away backend specific behavior and if it cannot do that
+it gets rid of some backend specific API. For example not all SYCL abstractions allow
+to get native backend handle which can be used to work with backend API directly.
+In general, this problem is solvable. There is an option to request Intel extensions
+which will provide required API
+
+For the models that have small execution time, for example DLRM or BERT the
+overhead becomes critical. All optimizations that can be done at oneDNN side is
+completed. They allowed to get rid of overhead for models like ResNet-50 (MB 1024),
+but they didn't solve the problem.
+
+The only viable option is to get rid of the overhead completely. The proposal is
+to switch to caching compiled kernel rather than their binary representations.
+
+## How to deal with the unloading libraries issue?
+
+In general, it's seem unlikely to run into the issue because:
+1. According to the SYCL integration design, frameworks store GPU device and context
+abstractions globally. If that was an issue they wouldn't have done that
+2. Benchdnn and Gtests have global engine which contains GPU device and context
+abstractions and no issues so far
+
+If it becomes an issue, the workaround will be clearing the primitive cache by
+calling  `dnnl::set_primitive_cache_capacity(0);` before exit on the frameworks' side.
+
+Also, SYCL is trying to handle such situations properly. For example on Linux,
+they use `__attribute__((destructor))` with low priority to make sure that the libraries
+that do not use the constructor with lower priority will be unloaded before the SYCL
+Runtime. [Here](https://github.com/intel/llvm/blob/sycl/sycl/doc/GlobalObjectsInRuntime.md)
+is the document that describes what they do to prevent such issues.
+
+## Benefits from caching compiled kernels
+
+There is a number of benefits:
+1. Eliminate overhead on creating kernels from binary and copying scales
+2. Significantly reduce code complexity
+    * No need in having resource for GPU
+    * No need in having complex abstractions that can contain either binary or kernel
+3. Everything can be stored in the primitive
```

Comment:
can you elaborate on that one?

Reply:
For example we cannot store memory because it's context dependent as well as other context dependent things. But now we can do that whiteout any limitations.